elegy/losses/huber.py

```diff
@@ -0,0 +1,186 @@
+from elegy import types
+import typing as tp
+
+import jax.numpy as jnp
+
+from elegy.losses.loss import Loss, Reduction
+
+
+def huber(y_true: jnp.ndarray, y_pred: jnp.ndarray, delta: float) -> jnp.ndarray:
+    """
+    Computes the Huber loss between labels and predictions.
+
+    ```python
+    For each value x in error = y_true - y_pred:
+    loss = 0.5 * x^2                  if |x| <= d
```

Comment:
You can go fancy with LATEX if you want, we are using a plugin for this :)

Reply:
Checkout: https://github.com/poets-ai/elegy/blob/014647a3849e8ce58efdba721d4278824b9e4a07/elegy/regularizers/global_l1l2.py#L18

Reply:
```
r"""
    Computes the huber errors between labels and predictions.

    $$
    \text{For each value x in error = y_true - y_pred:} \\
    \text{loss}= 
    \begin{cases}
    \ 0.5 \times x^2 ,& \text{if } |x|\leq d\\
    0.5 \times d^2 + d \times (|x| - d),              & \text{otherwise} 
    \end{cases} \\ where \, d \, is \, delta. See: https://en.wikipedia.org/ wiki/Huber\_loss
    $$
    ```
I tried adding this Latex But I'm getting some weird error while doing ```mkdocs serve```
```
ERROR   -  mkdocstrings.extension: not well-formed (invalid token): line 39, column 19 
ERROR   -  Error reading page 'api/losses/Huber.md': not well-formed (invalid token): line 39, column 19 
Traceback (most recent call last):
  File "/home/neurlap/.pyenv/versions/elegy-2/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/__main__.py", line 133, in serve_command
    serve.serve(
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/serve.py", line 141, in serve
    config = builder()
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/serve.py", line 136, in builder
    build(config, live_server=live_server, dirty=dirty)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/build.py", line 271, in build
    _populate_page(file.page, config, files, dirty)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/build.py", line 171, in _populate_page
    page.render(config, files)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/structure/pages.py", line 175, in render
    self.content = md.convert(self.markdown)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/core.py", line 263, in convert
    root = self.parser.parseDocument(self.lines).getroot()
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/blockparser.py", line 90, in parseDocument
    self.parseChunk(self.root, '\n'.join(lines))
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/blockparser.py", line 105, in parseChunk
    self.parseBlocks(parent, text.split('\n\n'))
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/blockparser.py", line 123, in parseBlocks
    if processor.run(parent, blocks) is not False:
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocstrings/extension.py", line 183, in run
    as_xml = XML(ENTITIES + rendered)
  File "/usr/lib/python3.8/xml/etree/ElementTree.py", line 1320, in XML
    parser.feed(text)
xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 39, column 19
(elegy-2) [neurlap@abhinav-swiftsf31541 elegy]$ mkdocs serve
INFO    -  Building documentation... 
INFO    -  Cleaning site directory 
ERROR   -  mkdocstrings.extension: not well-formed (invalid token): line 39, column 19 
ERROR   -  Error reading page 'api/losses/Huber.md': not well-formed (invalid token): line 39, column 19 
Traceback (most recent call last):
  File "/home/neurlap/.pyenv/versions/elegy-2/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/__main__.py", line 133, in serve_command
    serve.serve(
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/serve.py", line 141, in serve
    config = builder()
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/serve.py", line 136, in builder
    build(config, live_server=live_server, dirty=dirty)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/build.py", line 271, in build
    _populate_page(file.page, config, files, dirty)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/commands/build.py", line 171, in _populate_page
    page.render(config, files)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocs/structure/pages.py", line 175, in render
    self.content = md.convert(self.markdown)
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/core.py", line 263, in convert
    root = self.parser.parseDocument(self.lines).getroot()
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/blockparser.py", line 90, in parseDocument
    self.parseChunk(self.root, '\n'.join(lines))
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/blockparser.py", line 105, in parseChunk
    self.parseBlocks(parent, text.split('\n\n'))
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/markdown/blockparser.py", line 123, in parseBlocks
    if processor.run(parent, blocks) is not False:
  File "/home/neurlap/.pyenv/versions/elegy-2/lib/python3.8/site-packages/mkdocstrings/extension.py", line 183, in run
    as_xml = XML(ENTITIES + rendered)
  File "/usr/lib/python3.8/xml/etree/ElementTree.py", line 1320, in XML
    parser.feed(text)
xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 39, column 19
```

Reply:
@cgarciae Would you please help me with this PR. As you can see in the below picture I've written the formula in  the LaTex correctly and it's also rendering correctly.
But when I'm using the same LaTex insid the ```Huber``` docstring it's throwing me error while making the docs using mkdocs.
Latex:
```$$
\text{For each value x in error = y_true - y_pred:} \\
\text{loss}= 
\begin{cases}
\ 0.5 \times x^2 ,& \text{if } |x|\leq d\\
0.5 \times d^2 + d \times (|x| - d),              & \text{otherwise} 
\end{cases}
$$
```
Thanks.
![Screenshot_20201213_060642](https://user-images.githubusercontent.com/43638955/101992809-dbea4580-3cdb-11eb-99c3-bcbdfdadc8b6.png)

Reply:
This works butt does not align the right part:
```
$$
loss(x) =
\begin{cases}
\ 0.5 \times x^2 , \text{if } |x|\leq d\\
0.5 \times d^2 + d \times (|x| - d), \text{otherwise} 
\end{cases}
$$
```
It seems that the `&` character is not well received.

Reply:
@cgarciae  Thank you for fixing the LaTex error. I've update my PR kindly review again.